spring-core/src/main/java/org/springframework/util/ResourceUtils.java

```diff
@@ -384,4 +389,23 @@ public static void useCachesIfNecessary(URLConnection con) {
 		con.setUseCaches(con.getClass().getSimpleName().startsWith("JNLP"));
 	}
 
+	/**
+	 * Determine, if possible, the contentLength of the given resource
+	 * without reading it.
+	 * @param resource the resource instance
+	 * @return the contentLength of the resource
+	 */
+	public static OptionalLong contentLength(Resource resource) {
+		// Don't try to determine contentLength on InputStreamResource - cannot be read afterwards...
+		// Note: custom InputStreamResource subclasses could provide a pre-calculated content length!
+		if (InputStreamResource.class != resource.getClass()) {
+			try {
+				return OptionalLong.of(resource.contentLength());
+			}
+			catch (IOException ignored) {
+			}
+		}
+		return OptionalLong.empty();
+	}
+
```

Comment:
I don't mind an addition to `Resource`, in particular now where we can do default methods. However, I feel that we're overusing `Optional`. On such core abstractions, can we rather do it in a more traditional style through a `Long` return type which may simple be `null` if not known?

To be clear, using `Optional` in a functional API is perfectly sensible. I just don't want it to creep into Java interfaces at a lower level, in particular not in existing interfaces with a traditional design mindset in other methods... and not really in new interfaces in the same packages either.